File: goserver/output_config.go
package main

const (
	outputConfigSchemaVersion  = 1
	displayFontSizeMin         = 24
	displayFontSizeMax         = 96
	displayFontSizeDefault     = 48
	displayPanelOpacityMin     = 10
	displayPanelOpacityMax     = 100
	displayPanelOpacityDefault = 55
	blindBoxViewerSlotsMin     = 1
	blindBoxViewerSlotsMax     = 10
	blindBoxViewerSlotsDefault = 3
	defaultDisplayAccentColor  = "#fb7299"
)

var displaySceneLayoutIDs = map[string]struct{}{
	"stack": {}, "grid": {}, "focus": {}, "versus": {}, "dashboard": {},
}

var giftTargetLayoutIDs = map[string]struct{}{
	"stack": {}, "grid": {}, "dashboard": {},
}

func isDisplaySceneLayout(value string) bool {
	_, exists := displaySceneLayoutIDs[value]
	return exists
}

func normalizeDisplaySceneLayout(value string) string {
	if isDisplaySceneLayout(value) {
		return value
	}
	return "stack"
}

func isGiftTargetLayout(value string) bool {
	_, exists := giftTargetLayoutIDs[value]
	return exists
}

func normalizeGiftTargetLayout(value string) string {
	if isGiftTargetLayout(value) {
		return value
	}
	return "grid"
}

func displayAppearanceDefaults(settings settingsState) displayAppearanceState {
	return displayAppearanceState{
		ThemeID:        settings.DefaultDisplayThemeID,
		FontSize:       settings.FontSize,
		AccentColor:    settings.AccentColor,
		ShowConnection: settings.ShowConnection,
		Align:          settings.Align,
		PanelOpacity:   settings.PanelOpacity,
	}
}

func normalizeDisplayAppearanceState(appearance *displayAppearanceState, fallback displayAppearanceState) {
	if appearance == nil {
		return
	}
	if !isDisplayThemeID(appearance.ThemeID) {
		appearance.ThemeID = fallback.ThemeID
	}
	if appearance.FontSize <= 0 {
		appearance.FontSize = fallback.FontSize
	}
	appearance.FontSize = minInt(displayFontSizeMax, maxInt(displayFontSizeMin, appearance.FontSize))
	if !isHexColor(appearance.AccentColor) {
		appearance.AccentColor = fallback.AccentColor
	}
	if appearance.Align != "left" && appearance.Align != "center" && appearance.Align != "right" {
		appearance.Align = fallback.Align
	}
	if appearance.PanelOpacity <= 0 {
		appearance.PanelOpacity = fallback.PanelOpacity
	}
	appearance.PanelOpacity = minInt(displayPanelOpacityMax, maxInt(displayPanelOpacityMin, appearance.PanelOpacity))
}

func normalizeBlindBoxDisplayAppearanceState(appearance *blindBoxDisplayAppearanceState, fallback displayAppearanceState) {
	normalizeDisplayAppearanceState(&appearance.displayAppearanceState, fallback)
	if appearance.ViewerSlots <= 0 {
		appearance.ViewerSlots = blindBoxViewerSlotsDefault
	}
	appearance.ViewerSlots = minInt(blindBoxViewerSlotsMax, maxInt(blindBoxViewerSlotsMin, appearance.ViewerSlots))
}

func validateDisplayAppearanceState(appearance displayAppearanceState) bool {
	return isDisplayThemeID(appearance.ThemeID) &&
		isHexColor(appearance.AccentColor) &&
		appearance.FontSize >= displayFontSizeMin && appearance.FontSize <= displayFontSizeMax &&
		appearance.PanelOpacity >= displayPanelOpacityMin && appearance.PanelOpacity <= displayPanelOpacityMax &&
		(appearance.Align == "left" || appearance.Align == "center" || appearance.Align == "right")
}

File: src/output-config.ts
import { normalizeDisplayThemeId } from './display-themes';
import type {
  AppState,
  BlindBoxDisplayAppearance,
  DisplayAppearance,
  DisplaySceneLayout,
  DisplayThemeId,
  GiftKpiLayout,
} from './types';

export const OUTPUT_CONFIG_SCHEMA_VERSION = 1;
export const DISPLAY_SCENE_LAYOUT_IDS = ['stack', 'grid', 'focus', 'versus', 'dashboard'] as const satisfies readonly DisplaySceneLayout[];
export const GIFT_TARGET_LAYOUT_IDS = ['stack', 'grid', 'dashboard'] as const satisfies readonly GiftKpiLayout[];
export const DISPLAY_FONT_SIZE_RANGE = { min: 24, max: 96, default: 48 } as const;
export const DISPLAY_PANEL_OPACITY_RANGE = { min: 10, max: 100, default: 55 } as const;
export const BLIND_BOX_VIEWER_SLOTS_RANGE = { min: 1, max: 10, default: 3 } as const;
export const DEFAULT_DISPLAY_ACCENT_COLOR = '#fb7299';

export type DisplayAppearanceFallback = Pick<
  AppState['settings'],
  'fontSize' | 'accentColor' | 'showConnection' | 'align' | 'panelOpacity' | 'defaultDisplayThemeId'
>;

const displaySceneLayoutIds = new Set<string>(DISPLAY_SCENE_LAYOUT_IDS);
const giftTargetLayoutIds = new Set<string>(GIFT_TARGET_LAYOUT_IDS);

export function isDisplaySceneLayout(value: unknown): value is DisplaySceneLayout {
  return typeof value === 'string' && displaySceneLayoutIds.has(value);
}

export function normalizeDisplaySceneLayout(value: unknown): DisplaySceneLayout {
  return isDisplaySceneLayout(value) ? value : 'stack';
}

export function normalizeGiftTargetLayout(value: unknown): GiftKpiLayout {
  return typeof value === 'string' && giftTargetLayoutIds.has(value) ? value as GiftKpiLayout : 'grid';
}

export function normalizeDisplayAppearance(
  appearance: Partial<DisplayAppearance> | undefined,
  fallback: DisplayAppearanceFallback,
  fallbackThemeId: unknown = fallback.defaultDisplayThemeId,
): DisplayAppearance {
  const fontSize = finiteNumber(appearance?.fontSize, fallback.fontSize, DISPLAY_FONT_SIZE_RANGE.default);
  const panelOpacity = finiteNumber(appearance?.panelOpacity, fallback.panelOpacity, DISPLAY_PANEL_OPACITY_RANGE.default);
  const accentCandidate = String(appearance?.accentColor ?? fallback.accentColor ?? DEFAULT_DISPLAY_ACCENT_COLOR);
  const fallbackAlign = fallback.align === 'left' || fallback.align === 'right' ? fallback.align : 'center';
  const align = appearance?.align === 'left' || appearance?.align === 'right' || appearance?.align === 'center'
    ? appearance.align
    : fallbackAlign;
  return {
    themeId: normalizeDisplayThemeId(appearance?.themeId ?? fallbackThemeId) as DisplayThemeId,
    fontSize: clamp(fontSize, DISPLAY_FONT_SIZE_RANGE.min, DISPLAY_FONT_SIZE_RANGE.max),
    accentColor: /^#[0-9a-f]{6}$/i.test(accentCandidate) ? accentCandidate : DEFAULT_DISPLAY_ACCENT_COLOR,
    showConnection: appearance?.showConnection ?? fallback.showConnection,
    align,
    panelOpacity: clamp(panelOpacity, DISPLAY_PANEL_OPACITY_RANGE.min, DISPLAY_PANEL_OPACITY_RANGE.max),
  };
}

export function normalizeBlindBoxDisplayAppearance(
  appearance: Partial<BlindBoxDisplayAppearance> | undefined,
  fallback: DisplayAppearanceFallback,
): BlindBoxDisplayAppearance {
  const viewerSlots = finiteNumber(
    appearance?.viewerSlots,
    BLIND_BOX_VIEWER_SLOTS_RANGE.default,
    BLIND_BOX_VIEWER_SLOTS_RANGE.default,
  );
  return {
    ...normalizeDisplayAppearance(appearance, fallback),
    viewerSlots: clamp(Math.trunc(viewerSlots), BLIND_BOX_VIEWER_SLOTS_RANGE.min, BLIND_BOX_VIEWER_SLOTS_RANGE.max),
  };
}

function finiteNumber(value: unknown, fallback: unknown, finalFallback: number): number {
  const candidate = Number(value ?? fallback);
  if (Number.isFinite(candidate)) return candidate;
  const fallbackCandidate = Number(fallback);
  return Number.isFinite(fallbackCandidate) ? fallbackCandidate : finalFallback;
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}
